api/submission/fetch_course_report_test.go

```diff
@@ -0,0 +1,57 @@
+package submission
+
+import (
+    "reflect"
+    "testing"
+
+    "github.com/edulinq/autograder/api/core"
+    "github.com/edulinq/autograder/model"
+    "github.com/edulinq/autograder/report"
+    "github.com/edulinq/autograder/util"
+)
+
+func TestCourseReport(test *testing.T) {
+    testCases := []struct{
+        role model.UserRole
+        permError bool
+        expectedReport *FetchCourseReportResponse
+    }{
+        // Admin, Valid Permission
+        {model.RoleAdmin, false, &FetchCourseReportResponse{CourseReport: report.TestCourseReport}},
+
+        // Grader, Valid Permission
+        {model.RoleGrader, false, &FetchCourseReportResponse{CourseReport: report.TestCourseReport}},
+
+        // Student, Invalid Permission
+        {model.RoleStudent, true, nil},
+    };
+
+    for i, testCase := range testCases {
+        response := core.SendTestAPIRequestFull(test, core.NewEndpoint(`submission/fetch/course-report`), nil, nil, testCase.role);
+        if (!response.Success) {
+            if (testCase.permError) {
+                expectedLocator := "-020";
+                if (response.Locator != expectedLocator) {
+                    test.Errorf("Case %d: Incorrect error returned on permissions error. Expcted '%s', found '%s'.",
+                        i, expectedLocator, response.Locator);
+                }
+            } else {
+                test.Errorf("Case %d: Response is not a success when it should be: '%v'.", i, response);
+            }
+
+            continue;
+        }
+
+        var responseContent *FetchCourseReportResponse;
+        util.MustJSONFromString(util.MustToJSON(response.Content), &responseContent);
+
+        var expectedResponse *FetchCourseReportResponse;
```

Comment:
No need to serialize and deserialize the expected result.
With the response, we don't have a strict type on it, so we need to convert it, but you already have the expected type.

Reply:
I marshaled the expected because of the formatting info that comes with the expected test from the report/test.go. Initially, the test was written for course_test.go; it has formatting fields because of the html tests. If I don't serialize and deserialize the test, the DeepEqulas will fail because of the formatting fields.

Reply:
Okay that makes sense.
Add a comment so people will know.